### DISCUSSION TITLE
Adding Transaction Support as a Middle-Step to Supporting User Groups

### DISCUSSION THREAD

**Discussion by Uriah372-DS** [fastapi-users/fastapi-users#1445] [Ideas]
Hello!

Let me start by thanking everyone for this incredibly useful tool, really amazing and lightweight, and also strikes the perfect balance between encapsulating the complexity of authentication and giving control to the developer.

**TL;DR:** Consider adding support for transactional operations to the database adapters before adding User Groups feature.

I read in your docs sometime ago that you're planning to add support for User Groups / Organizations Management. I was searching for something like this and been trying to extend this library to add this feature on my own, however I've run into a wall when trying to create a User+Organization Registration flow. 

Specifically, I copied the full example that uses the Beanie adapter and tried implementing a registration route that registers both a user and an organization (e.g., for B2B applications). 

However, this operation ended up consisting of multiple steps that need to be performed against the database:
1. Create a User and an Organization (and optionally default roles, permissions, etc.) instances
2. Upload them to the database
3. Link the instances to each-other and update them in the database (can only be done after they each have an ObjectId).

To ensure consistency, these operations should be done in a transaction, however it seems that transactions aren't supported in the BeanieUserDatabase adapter.

I'd be happy to know what you think about this, if you disagree that this step is necessary and if so then why, and if adding support for user groups is still on the table.

Thanks again!

**Comment by hgalytoby** [comment]
The same goes for sqlalchemy. If you need multiple operations to be done in a transaction, you may need to write a new api function.
